content/recipes/authoring/kata-snippets/submission-tests.md

```diff
@@ -0,0 +1,88 @@
+- Under some rare circumstances, it's allowed to use so called [**randomized tests**](/glossary/#randomized-tests) instead of fully random ones. For some types of problems (for example, kata simulating chess, or problems with small set of possible inputs) it may be too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too). Before the tests are run, the set of inputs should be shuffled, and eventually randomly sampled at each run to make the hard-coding of the results more tedious for the user. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.
+- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
+- **Difficulty should be consistent between test runs.** When randomized ranges are very large, it becomes possible for some users to receive many small, easy inputs while other users receive the exact opposite. Make sure that your random tests are built in a way which minimizes the chances for such a situation to occur. If you want to test difficult inputs, split the test cases into a set of easy ones and a set of difficult ones, and test them separately.
+- If it's not possible to build fixed tests in a way they'd be able to catch most of the failures and you have to rely on random tests to catch errors, **make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
+
+## Performance Tests
+
+Some kata require solutions to be fast enough. For example, the author may only wish to accept solutions completing in (sub-)linear time. Building such test suites is not an easy task!
+
+- **Performance tests can be implemented in terms of random tests by testing with large, random inputs.** However, to make debugging easier, it can be worth to have a separate set of random tests with small inputs.
+- If the difficulty of a kata is roughly proportional to the size of the input, it's usually **better to have a few tests with large input rather than many with small inputs.** For example, 200 tests with huge numbers or arrays is usually better than 10,000 tests with small ones (but see remarks on size of inputs above).
```

Comment:
```suggestion
- If the difficulty of a kata is roughly proportional to the size of the input, it's usually **better to have a few tests with large input rather than many with middle sized inputs.** For example, 100 tests with huge numbers or arrays is usually better than 1000 tests with arrays of size 10000 (but see remarks on size of inputs above and the problem below about building the inputs too).
```

Reply:
OK, but "1000 tests with arrays of size 10000" seems oddly specific to me ... ;-)

Reply:
yeah, it was just to give an idea, but it's so much depending on the language... Update it as you prefer.

Reply:
Done